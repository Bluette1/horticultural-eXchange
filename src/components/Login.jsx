import React, { useState, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import Form from 'react-validation/build/form';
import Input from 'react-validation/build/input';
import CheckButton from 'react-validation/build/button';
import GuestLogin from './GuestLogin';
import AuthService from '../services/auth.service';
import { loginSuccess, loginFail, setMessage } from '../actions/auth';

const required = (value) => {
  if (!value) {
    return (
      <div className="alert alert-danger" role="alert">
        This field is required!
      </div>
    );
  }
  return null;
};

const Login = ({ history }) => {
  const form = useRef();
  const checkBtn = useRef();

  const [guest, setGuest] = useState(false);
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [loading, setLoading] = useState(false);

  const { isLoggedIn } = useSelector((state) => state.auth);
  const { message } = useSelector((state) => state.message);

  const dispatch = useDispatch();

  const onChangeEmail = (e) => {
    const email = e.target.value;
    setEmail(email);
  };

  const onChangePassword = (e) => {
    const password = e.target.value;
    setPassword(password);
  };
  const handleGuestLogin = (e) => {
    e.preventDefault();
    setGuest(true);
  };

  const handleLogin = (e) => {
    e.preventDefault();

    setLoading(true);

    form.current.validateAll();

    if (checkBtn.current.context._errors.length === 0) {
      AuthService.login(email, password)
        .then((response) => {
          const user = response.data;

          if (response.headers.authorization) {
            user.accessToken = response.headers.authorization;
          }

          localStorage.setItem('user', JSON.stringify(user));

          dispatch(loginSuccess(user));

          history.push('/');
          window.location.reload();
        })
        .catch((error) => {
          const message = (error.response
            && error.response.data
            && error.response.data.message)
            || error.message
            || error.toString();

          dispatch(loginFail());
          dispatch(setMessage(message));
          setLoading(false);
        });
    } else {
      setLoading(false);
    }
  };

  if (isLoggedIn) {
    return <Redirect to="/profile" />;
  }

  if (guest) {
    return <GuestLogin history={history} />;
  }

  return (
    <div className="col-md-12" data-testid="login-container">
      <div className="card card-container">
        <img
          src="//ssl.gstatic.com/accounts/ui/avatar_2x.png"
          alt="profile-img"
          className="profile-img-card"
        />

        <Form onSubmit={handleLogin} ref={form}>
          <div className="form-group">
            <p>email</p>
            <Input
              type="text"
              className="form-control"
              name="email"
              value={email}
              onChange={onChangeEmail}
              validations={[required]}
            />
          </div>

          <div className="form-group">
            <p>Password</p>
            <Input
              type="password"
              className="form-control"
              name="password"
              value={password}
              onChange={onChangePassword}
              validations={[required]}
            />
          </div>

          <div className="form-group mt-3">
            <button
              className="btn btn-primary btn-block"
              disabled={loading}
              type="submit"
              data-testid="submit-btn"
            >
              {loading && <span className="spinner-border spinner-border-sm" />}
              <span>Login</span>
            </button>
          </div>

          {message && (
            <div className="form-group">
              <div className="alert alert-danger" role="alert">
                {message}
              </div>
            </div>
          )}
          <CheckButton style={{ display: 'none' }} ref={checkBtn} />
        </Form>
      </div>
      <div className="d-flex justify-content-end">
        <div className="col-md-4">
          <span
            style={{
              color: '#008000',
              fontWeight: 'bold',
              padding: '5px',
              border: '1px solid',
              cursor: 'pointer',
            }}
            role="presentation"
            onKeyDown={handleGuestLogin}
            onClick={handleGuestLogin}
            data-testid="guest-login"
          >
            Guest Login
          </span>
        </div>
      </div>
    </div>
  );
};

Login.propTypes = {
  history: PropTypes.objectOf(PropTypes.any).isRequired,
};

export default Login;
