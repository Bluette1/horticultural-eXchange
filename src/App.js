import React, { useState, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Router, Switch, Route, Link } from "react-router-dom";
import "bootstrap/dist/css/bootstrap.min.css";
import "./App.css";
import logo from "./logo.png";
import Login from "./components/Login";
import Register from "./components/Register";
import Home from "./components/Home";
import Profile from "./components/Profile";
import StripeLoader from "./components/StripeLoader";
import Cart from "./components/Cart";
import BoardUser from "./components/BoardUser";
import BoardModerator from "./components/BoardModerator";
import BoardAdmin from "./components/BoardAdmin";
import PlantForm from './components/PlantForm';
import CategoryForm from './components/CategoryForm';
import Category from './components/Category';
import { logout } from "./actions/auth";
import { clearMessage } from "./actions/message";

import { history } from "./helpers/history";

const App = () => {
  const [showModeratorBoard, setShowModeratorBoard] = useState(false);
  const [showAdminBoard, setShowAdminBoard] = useState(false);

  const { user: currentUser } = useSelector((state) => state.auth);
  const cartItems = useSelector(state => state.cart);

  const dispatch = useDispatch();
  const { message } = useSelector((state) => state.message);

  useEffect(() => {
    history.listen(() => {
      dispatch(clearMessage()); // clear message when changing location
    });
  }, [dispatch]);

  useEffect(() => {
    if (currentUser) {
      setShowModeratorBoard(currentUser.supervisor_role);
      setShowAdminBoard(currentUser.superadmin_role);
    }
  }, [currentUser]);

  const logOut = (e) => {
    e.preventDefault();
    dispatch(logout())
  };

  return (
    <Router history={history}>
      <div>
        <nav className="navbar navbar-expand navbar-dark bg-dark d-flex justify-content-around justify-content-lg-between">
        <div>
       
          <div className="navbar-nav mr-auto">
          <Link to={"/"} className="navbar-brand">
            <img
              style={{
                marginRight: "2px",
                width: "30px",
                height: "30px",
                borderRadius: 50,
              }}
              src={logo}
              alt=""
            />{" "}
            XChange
          </Link>
            <li className="nav-item">
              <Link to={"/home"} className="nav-link">
                Home
              </Link>
            </li>

            {showModeratorBoard && (
              <li className="nav-item">
                <Link to={"/mod"} className="nav-link">
                  Moderator Board
                </Link>
              </li>
            )}

            {showAdminBoard && (
              <li className="nav-item">
                <Link to={"/admin"} className="nav-link">
                  Admin Board
                </Link>
              </li>
            )}

            {currentUser && (
              <li className="nav-item">
                <Link to={"/user"} className="nav-link">
                  User
                </Link>
              </li>
              
            )}
            {currentUser && (<li className="nav-item">
              <Link to={"/profile"} className="nav-link">
                {currentUser.email}
              </Link>
            </li>)}
            
          </div>
        </div>

          <div>{currentUser ? (
            <div className="navbar-nav ml-auto">
              <li className="nav-item d-flex">
                <a href="/login" className="nav-link" onClick={logOut}>
                  Logout
                </a>
                {message && (
                  <span className="alert alert-danger">
                    {message}
                  </span>
                )}
              </li>
              <li className="nav-item d-flex">
              <Link to={"/cart"} className="nav-link">
              <i class="fa fa-shopping-cart" aria-hidden="true"></i>
              <span style={{
                borderRadius: "40%", backgroundColor: "#9d9d9d", color: "white", fontSize: "8px",
                padding: "2px", marginRight: "3px",
                }}>{cartItems.length}</span>
                </Link>
              </li>
            </div>
          ) : (
            <div className="navbar-nav ml-auto">
              <li className="nav-item">
                <Link to={"/login"} className="nav-link">
                  Login
                </Link>
              </li>

              <li className="nav-item">
                <Link to={"/register"} className="nav-link">
                  Sign Up
                </Link>
              </li>
            </div>
          )}</div>
        </nav>

        <div className="container mt-3">
          <Switch>
            <Route exact path={["/", "/home"]} component={Home} />
            <Route exact path="/login" component={Login} />
            <Route exact path="/register" component={Register} />
            <Route exact path="/profile" component={Profile} />
            <Route exact path="/payment" component={StripeLoader} />
            <Route exact path="/cart" component={Cart} />
            <Route exact path="/new-plant" component={PlantForm} />
            <Route exact path="/new-category" component={CategoryForm} />
            <Route path="/user" component={BoardUser} />
            <Route path="/mod" component={BoardModerator} />
            <Route path="/admin" component={BoardAdmin} />
            <Route path="/category/" component={Category} />
          </Switch>
        </div>
      </div>
    </Router>
  );
};

export default App;
