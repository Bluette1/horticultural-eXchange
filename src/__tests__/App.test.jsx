import React from 'react';
import {
  render, within, cleanup, waitFor,
} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Provider } from 'react-redux';
import axios from 'axios';
import App from '../App';
import store from '../store';
import { httpProtocol, host, port } from '../env.variables';
import Login from '../components/Login';

jest.mock('axios');

afterEach(cleanup);

const AppWithStore = () => (
  <Provider store={store}>
    <React.StrictMode>
      <App />
    </React.StrictMode>
  </Provider>
);
test('renders the app - navbar links are displayed correctly', async () => {
  axios.get.mockImplementation((url) => {
    switch (url) {
      case `${httpProtocol}://${host}:${port}/categories`:
        return Promise.resolve({ data: [] });
      case `${httpProtocol}://${host}:${port}/plants`:
        return Promise.resolve({ data: [] });
      default:
        return Promise.resolve({ data: [] });
    }
  });
  const { container } = render(<AppWithStore />);
  await waitFor(() => {
    // const navBar = container.querySelector('.navbar');
    const navBar = container.firstChild.firstChild;
    expect(navBar).toHaveClass('navbar navbar-expand-lg navbar-dark bg-navbar d-flex justify-content-around justify-content-lg-between');
    const logoImg = within(navBar).getByAltText("logo");
    expect(logoImg).toBeInTheDocument();
    expect(logoImg).toHaveAttribute('src', 'logo.png');
    expect(within(navBar).getByText(/iGrow/)).toBeInTheDocument();

    expect(navBar).toBeTruthy();
    expect(navBar.children).toHaveLength(2);

    expect(container).toMatchSnapshot();

    const navItems = navBar.querySelectorAll('.nav-item');
    expect(navItems).toHaveLength(3);
    
    const home = navItems[0].firstChild;
    expect(within(home).getByText("Home")).toBeInTheDocument();
    expect(home).toContainHTML('<a class="nav-link" href="/home">Home</a>');

    const login = navItems[1].firstChild;
    expect(within(login).getByText("Login")).toBeInTheDocument();
    expect(login).toContainHTML('<a class="nav-link" href="/login">Login</a>');

    const signup = navItems[2].firstChild;
    expect(within(signup).getByText("Sign Up")).toBeInTheDocument();
    expect(signup).toContainHTML('<a class="nav-link" href="/register">Sign Up</a>');
  });
});