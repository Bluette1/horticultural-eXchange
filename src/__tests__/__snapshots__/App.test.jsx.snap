// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the app - navbar links are displayed correctly 1`] = `
<div>
  <div>
    <nav
      class="navbar navbar-expand-lg navbar-dark bg-navbar d-flex justify-content-around justify-content-lg-between"
    >
      <div>
        <button
          aria-controls="navbarToggler03"
          aria-expanded="false"
          aria-label="Toggle navigation"
          class="navbar-toggler"
          data-bs-target="#navbarToggler03"
          data-bs-toggle="collapse"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="navbar-nav mr-auto collapse navbar-collapse"
          id="navbarToggler03"
        >
          <a
            class="navbar-brand"
            href="/"
          >
            <img
              alt="logo"
              src="logo.png"
              style="margin-right: 2px; width: 30px; height: 30px; border-radius: 50px;"
            />
             
            iGrow
          </a>
          <li
            class="nav-item"
          >
            <a
              class="nav-link"
              href="/home"
            >
              Home
            </a>
          </li>
        </div>
      </div>
      <div>
        <button
          aria-controls="navbarToggler04"
          aria-expanded="false"
          aria-label="Toggle navigation"
          class="navbar-toggler"
          data-bs-target="#navbarToggler04"
          data-bs-toggle="collapse"
          type="button"
        >
          <i
            aria-hidden="true"
            class="fa fa-user toggler-icon"
          />
        </button>
        <div
          class="navbar-nav ml-auto collapse navbar-collapse"
          id="navbarToggler04"
        >
          <div
            class="d-lg-flex"
          >
            <li
              class="nav-item"
            >
              <a
                class="nav-link"
                href="/login"
              >
                Login
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link"
                href="/register"
              >
                Sign Up
              </a>
            </li>
          </div>
        </div>
      </div>
    </nav>
    <div
      class="container mt-3"
    >
      <div
        class="container row d-flex"
      >
        <h4>
          PRODUCT CATEGORIES:
        </h4>
        <div
          class="col-md-3"
        >
          <label
            for="category-select"
          >
            PRODUCT CATEGORIES:
            <select
              id="category-select"
              name="categories"
            >
              <option
                value=""
              >
                 Select a category
              </option>
              <option
                value="All Plants"
              >
                All Plants
              </option>
            </select>
          </label>
        </div>
        <div
          class="col-md-9"
        >
          <div
            class="row d-flex"
          >
             
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
